Posts: JWT signed streams (#22)
anycable/anycable-web: src/blog/jwt-identification-and-hot-streams/index.md

# AnyCable PRO:<br>JWT identification and "hot streams"

September 27, 2021
{date}

> When we started collecting applications for the [AnyCable PRO][pro] early access program back in July, we asked respondents to share the way they use _cables_ and which features they would like us to provide out-of-the-box. The analysis of their responses led to the development of two new PRO features, [JWT identification][jwt-id-docs] and [Signed streams support][signed-streams-docs], which might seem unrelated to each other from the first look but turned out to bring a huge performance boost when used together. Continue reading to learn more!

Reviewer: "... and which features they would like us to provide out of the box. Analyzing these responses led to the development of two new PRO features, [JWT identification][jwt-id-docs] and [signed streams support][signed-streams-docs]. They might seem unrelated to each other at first glance, but when used together, they have the potential to bring about a huge performance boost. Continue reading to learn more!"

{intro}

<div class="divider"></div>

## Dealing with authentication

> Among the requested features, one of the leaders was "a token-based authentication". Most Action Cable applications relies on cookies as an authentication mechanism (directly or via sessions). The main benefit of the cookie-based authentication is the simplicity: it just works. However, there are some drawbacks:

Reviewer: Amongst all the asked about features, one of the leading requests was a "token-based authentication". Most Action Cable applications rely on cookies as an authentication mechanism (either directly or via sessions). The main benefit of cookie-based authentication is simplicity–it just works. However, there are some drawbacks:

> - Authenticating non-web clients (e.g., mobile apps) is becoming cumbersome.

Reviewer: "is becoming" or "has become"? The thing is, "is becoming" sounds odd.

Author: "Becomes"?

Reviewer: No, it cannot be "becomes", that would be like the feeling in Japanese poem, if that makes any sense.

The primary question is -- is this still happening, or has it passed the point of no return -- is it cumbersome already, or is it in the process of becoming cumbersome?

Author: What I wanted to say here is that authenticating mobile clients via cookies is complicated.

So, yeah, it's probably just "is", there is no process or whatever.

Reviewer: yep, just delete "becoming" then

Author: ```suggestion
- Authenticating non-web clients (e.g., mobile apps) is cumbersome.
```

> - Rails API-only apps usually do not use tokens for web clients authentication as well (they use tokens).

Reviewer: The meaning was lost here. It says they usually do not use tokens, then it says "they use tokens". Could use clarification here.

Author: ```suggestion
- Rails API-only apps usually do not use cookies for web clients authentication as well (they use tokens).
```

> - WebSockets have no CORS support and, thus, **vulnerable to cross-site request forgery** (or [cross-site WebSocket hijacking][cross-site-ws-hijack]).

Reviewer: WebSockets do not offer CORS support and, thus, are **vulnerable to cross-site request forgery** (or [cross-site WebSocket hijacking][cross-site-ws-hijack]).

> Using tokens could help us to solve these problems. What's the cost? Well, we have to build everything ourselves, both server and client side (and good luck with finding a library or gem which makes it easier not harder).

Reviewer: Using tokens could certainly help us solve these problems–but what's the catch? Well, we'd have to build everything ourselves, both the server and the client side (and good luck actually finding a library or gem which makes this process easier instead of harder).

> While I was thinking on how to incorporate tokens into AnyCable, I realized that we can also leverage this feature to improve the performance: tokens could also carry **identification** information, and thus, could be used instead of calling an RPC server (`Authenticate` method).

Reviewer: While I was thinking about how to incorporate tokens into AnyCable, I realized that we could also leverage this feature to improve performance. Tokens could also be made to carry **identification** information, and therefore could be used instead of calling an RPC server (`Authenticate` method).

Reviewer: @palkan probably needs a review -- specifically the phrase "could also be made to carry"

> Let's recall how Action Cable (and AnyCable) works in the context of authentication and identification.

Reviewer: works --> work

Author: ```suggestion
Let's recall how Action Cable (and AnyCable) work in the context of authentication and identification.
```

> Whenever a new connection opens, the `ApplicationCable::Connection#connect` method is called. That's the place where you can reject the connection ([`#reject_unathorized_connection`][reject-connection]) and where you configure so-called _connection identifiers_ ([`.identifed_by`][ac-identifiers]). Identifiers represent the client state which could be used by channels (e.g., `current_user` or `current_tenant`). AnyCable obtains identifiers during the `Authenticate` call and pass along all subsequent requests (in a [serialized form][identifiers-serialization]).

Reviewer: Whenever a new connection opens, the `ApplicationCable::Connection#connect` method is called. This is the place where you can reject the connection ([`#reject_unathorized_connection`][reject-connection]) and where you configure the so-called _connection identifiers_ ([`.identifed_by`][ac-identifiers]). The identifiers represent the client state which could be used by channels (e.g., `current_user` or `current_tenant`). AnyCable obtains the identifiers during the `Authenticate` call and passes along all subsequent requests (in a [serialized form][identifiers-serialization]).

> To sum up, all we need from the `#connect` method is to accept-or-reject connection and populate identifiers. And we can encapsulate this logic within a single [JWT][] token!

Reviewer: To sum everything up, all we need from the `#connect` method is to accept-or-reject the connection and populate the identifiers. And we can encapsulate all this logic within a single [JWT][] token!

> To sum up, all we need from the `#connect` method is to accept-or-reject connection and populate identifiers. And we can encapsulate this logic within a single [JWT][] token!

Reviewer: @palkan probably needs a review here

> This is how it could look in pure Ruby. First, let's create a helper to build a connection url with a token:

Reviewer: This is how it might look written in pure Ruby. First, let's create a helper to build a connection URL with a token:

```ruby
class ApplicationHelper
  def action_cable_with_jwt_meta_tag(**identifiers)
    # Token TTL (5 minutes)
    exp = Time.now.to_i + 60*5
    # JWT payload.
    # The #serialize_identifiers is responsible for serializing objects into strings
    # (AnyCable uses GlobalID for that)
    payload = {ext: serialize_identifiers(identifiers), exp: exp}

    token = JWT.encode payload, JWT_ENCRYPTION_KEY, "HS256"

    base_url = ActionCable.server.config.url || ActionCable.server.config.mount_path
    tag "meta", name: "action-cable-url", content: "#{base_url}?token=#{token}"
  end
end
```

Now we can add it to our layout:

```erb
<%= action_cable_with_jwt_meta_tag(current_user: current_user) %> would render:
# => <meta name="action-cable-url" content="ws://demo.anycable.io/cable?token=eyJhbGciOiJIUzI1NiJ9....EWCEzziOx3sKyMoNzBt20a3QvhEdxJXCXaZsA-f-UzU" />
```

> Finally, we need to verify the token and extract the identifiers in the `#connect` method:

Reviewer: @palkan can we say "using" the connect method?

Author: The `#connect` method is always present; it's what's inside that matters. So, I think, "in" (or "at", or whatever).

Reviewer: OK

```ruby
module ApplicationCable
  class Connection < ActionCable::Connection::Base
    identified_by :current_user

    def connect
      token = request.params[:token]

      decoded_token = JWT.decode token, JWT_ENCRYPTION_KEY, true, { algorithm: "HS256" }
      # The #deserialize_identifiers method is a counterpart to #serialize_identifiers
      # and is responsible for converting strings back to objects
      identifiers = deserialize_identifiers(JSON.parse(decoded_token["ext"]))

      # Populate identifiers
      identifiers.each do |k, v|
        self.public_send("#{k}=", v)
      end
    rescue JWT::DecodeError
      reject_unauthorized_connection
    end
  end
end
```

> Doesn't look like a simple solution 😕 If you'll continue staring at this piece of code for a while, you may notice that the implementation has nothing specific to a particular application: we retrieve a token, verify it and extract identifiers. If so, why can't we re-implement this in Go? We can!

Reviewer: This doesn't look like such a simple solution, after all 😕 If you continue staring at this piece of code for a bit longer, you might notice that this implementation contains nothing specific to a particular application: we retrieve a token, verify it, and extract the identifiers. If this is the case, why can't we just re-implement this in Go? Well, we can!

This is how AnyCable Go PRO JWT identification works:

> - A client passes a token along the connection request (via a query string or a header).

Reviewer: along the --> along with the

Author: ```suggestion
- A client passes a token along with the connection request (via a query string or a header).
```

> - Go server retrieves a header, verifies its signature and TTL, and rejects the connection if token is invalid.

Reviewer: A Go server retrieves a header, verifies its signature and TTL, and rejects the connection if the token is invalid.

> - If token is valid, we extract identifiers, store them in the connection metadata and finalize the handshake (by sending a `{"type":"welcome"}` message).

Reviewer: If the token is valid, we extract the identifiers, store them in the connection metadata and finalize the handshake (by sending a `{"type":"welcome"}` message).

And we do not perform any RPC calls! Let's see how it affects the connection time:

<figure class="blog--figure">
<video loop="loop" muted="muted" autoplay="autoplay" playsinline="true" width="1221" height="800" class="blog--media">
  <source src="./connect-benchmark.h264.mp4" ttype="video/mp4; codecs=avc1.4D401E,mp4a.40.2">
</video>
</figure>

> More than 2x faster! What's more important, we reduce the stress on the RPC server. That could be especially useful during _connection avalanches_.

Reviewer: It's more than 2x faster now! And even more importantly, we've reduced the stress on the RPC server. This could be especially useful during _connection avalanches_.

> > JWT identification allows you to standardize the authentication flow for WebSockets, saves from cross-site WebSocket hijacking and brings a performance boost!

Reviewer: brings a performance boost! --> results in a performance boost!

> > JWT identification allows you to standardize the authentication flow for WebSockets, saves from cross-site WebSocket hijacking and brings a performance boost!

Reviewer: @palkan "saves" won't work here. Can we say "protects"?

Author: ```suggestion
> JWT identification allows you to standardize the authentication flow for WebSockets, protects from cross-site WebSocket hijacking and results in a performance boost!
```

> And that's what you can get with AnyCable PRO and a little companion gem, [anycable-rails-jwt][], which adds helpers to generate tokens.

Reviewer: @palkan This idea and sentence seem disconnected – would probably good to add some context about how this fits with the rest of the text. the phrase "and that's what you get" especially doesn't seem to work

> The only question left is how to handle tokens expiration gracefully? AnyCable Go uses a specific disconnect reason to distinguish expiration from unauthorized access (sends a `{"type":"disconnect","reason":"token_expired"}` message). You can use it to refresh the token. We plan to add a built-in refresh functionality to our [anycable-client][] library soon. Stay tuned!

Reviewer: The only remaining question–how to gracefully handle tokens expiration? AnyCable Go uses a specific disconnect reason to distinguish expiration from unauthorized access (it sends a `{"type":"disconnect","reason":"token_expired"}` message). You can use it to refresh the token. We plan to add a built-in refresh functionality to our [anycable-client][] library soon, so stay tuned!

Reviewer: @palkan "a specific disconnect reason" doesn't seem to work here. I see that "reason" is a key, is it a header? we need to use a different word here like, header, or property, I think. otherwise it will just appear to be a grammatical mistake.

Author: ```suggestion
The only remaining question–how to gracefully handle tokens expiration? AnyCable Go uses a specific disconnection message to distinguish expiration from unauthorized access (it sends a `{"type":"disconnect","reason":"token_expired"}` message). You can use it to refresh the token. We plan to add a built-in refresh functionality to our [anycable-client][] library soon, so stay tuned!
```

What about "disconnection message"?

> ## Adding some "hot wires" to the equation

Reviewer: @palkan Is it like a joke, like "hot wiring" a car? it's in quotes like a pun

Author: I referred to "Hotwire"  (https://hotwired.dev).

But, actually, the "hot wiring" also makes sense (subscribing to a channel without an RPC call => starting a car without a key), and I like it.

Reviewer: We basically have to say like this:
"Hot-wiring" the _______ 

Something must go there

Author: What about:

"Hot-wiring" Action Cable subscriptions

?

Reviewer: Yep! Sounds good.

Author: ```suggestion
## "Hot-wiring" Action Cable subscriptions
```

> We've collected more than a hundred responses from Action Cable / AnyCable users around the world, and found that a good portion of them build applications on top of [Hotwire][hotwire] or [Stimulus Reflex][sr] (+ [CableReady][cr]). Which is pretty cool: [_frontendless_ Rails frontend][frontendless-rails] is getting more and more traction!

Reviewer: We've collected more than a hundred responses from Action Cable and AnyCable users around the world, and we found that a good portion of them build applications on top of [Hotwire][hotwire] or [Stimulus Reflex][sr] (+ [CableReady][cr]). This is pretty cool: [_frontendless_ Rails frontend][frontendless-rails] is gaining more and more traction!

> What can AnyCable give to these users? Let's take a look at how Turbo Streams work with Rails.

Reviewer: give to these users --> do for these users

Author: ```suggestion
What can AnyCable do for these users? Let's take a look at how Turbo Streams work with Rails.
```

> We have the [turbo-rails][] gem, which integrate Action Cable with Hotwire. All you need is to drop a helper into your template to start consuming streams:

Reviewer: We'll start with the [turbo-rails][] gem, which integrates Action Cable with Hotwire. All you need to do is to drop a helper into your template in order to start consuming streams:

Reviewer: Let's start with the [turbo-rails][] gem, which integrates Action Cable with Hotwire. You just need to drop a helper into your template to start consuming streams:

```erb
<%= turbo_stream_from some_model %>
```

The helper adds a [`<turbo-cable-stream-source>` element][turbo-cable-stream-source], which initiates an Action Cable subscription. The corresponding server-side code is as follows:

```ruby
class Turbo::StreamsChannel < ActionCable::Channel::Base
  extend Turbo::Streams::Broadcasts, Turbo::Streams::StreamName

  def subscribed
    if verified_stream_name = self.class.verified_stream_name(params[:signed_stream_name])
      stream_from verified_stream_name
    else
      reject
    end
  end
end
```

The `#verified_stream_name` method uses an [`ActiveSupport::MessageVerifier`][message-verifier] to decode and verify the stream name. The subscription is rejected if the signed stream name is invalid.

> The code above is structurely the same as the one we wrote for JWT identification. What does it mean? We can "move" the implementation to AnyCable Go server and **create subscriptions without touching RPC**. Again.

Reviewer: Structurally speaking, the code above is the same as the example we wrote for JWT identification. What does this mean? It means that, **once again**, we can "move" the implementation to an AnyCable Go server and **create subscriptions without touching RPC**.

> That's exactly what we did as a part of the [Signed streams][signed-streams-docs] feature. And we did it not only for Turbo Streams, but for Cable Ready as well (since its [`#stream_from` implementation][cr-stream-from] is almost the same).

Reviewer: This is exactly what we did as a part of the [signed streams][signed-streams-docs] feature. Further, we did it not only for Turbo Streams, but for Cable Ready as well (since its [`#stream_from` implementation][cr-stream-from] is nearly the same).

> > Combining JWT identification with signed streams, it is possible to avoid running RPC server at all in case you only use Hotwire/CableReady functionality.

Reviewer: By combining JWT identification with signed streams, it's possible to completely avoid running an RPC server, in case you only need to use Hotwire/CableReady functionality.

Reviewer: @palkan i think it's better to avoid "/" here. can we use "and" or "or" instead?

> Below is the visualization of the RPC server metrics during Hotwire benchmarks with four different configuration (blue markers), from left to rifght: baseline (AnyCable PRO w/o any features enabled), with JWT identification, with signed streams, and with both features enabled at the same time.

Reviewer: Below is a visualization of RPC server metrics during Hotwire benchmarks with four different configurations (denoted by blue markers), from left to right: baseline (AnyCable PRO without any features enabled), with JWT identification, with signed streams, and with both features enabled at the same time.

<figure class="blog--figure">
  <img class="blog--media" title="AnyCable metrics when running with and without JWT and 'hot' streams" src="./grafana.png" width="1364" height="624">
</figure>

> No surpises that the last one shows zeros for RPC metrics—it hasn't been bothered at all. We can also notice that the AnyCable Go CPU usage is the lowest for the fourth scenario—**dealing with tokens and signatures is "cheaper" than performing gRPC calls**.

Reviewer: It shouldn't come as any surprise that the last one shows zeros for RPC metrics—it wasn't impacted at all. We can also note that CPU usage with AnyCable Go is the lowest in the fourth scenario—**dealing with tokens and signatures is "cheaper" than performing gRPC calls**.

<div class="divider"></div>

> We added these features in response to our users needs. because we believe that together we can build a much better product. Want to join The Team? Give [AnyCable a try today](/)!

Reviewer: I'm confused about this line here -- is this a pitch for AnyCable or to join Evil Martians? It's a CTA to join "The Team"

Author: By The Team I meant AnyCable users (with whom "we can build a much better product").

> We added these features in response to our users needs. because we believe that together we can build a much better product. Want to join The Team? Give [AnyCable a try today](/)!

Reviewer: The Team --> the team

# AnyCable PRO:<br>JWT identification and "hot streams"

September 27, 2021
{date}

> When we started collecting applications for the [AnyCable PRO][pro] early access program back in July, we asked respondents to share how they use _cables_ and which features they would like us to provide out of the box. Analyzing these responses led to the development of two new PRO features, [JWT identification][jwt-id-docs] and [signed streams support][signed-streams-docs]. They might seem unrelated to each other at first glance, but when used together, they have the potential to bring about a huge performance boost. Continue reading to learn more!

Reviewer 2: _I suggest rewriting like that (to explain to newcomers from Ruby Weekly that's the product). Feel free to correct/improve this piece._

We've just launched the final [AnyCable PRO][pro] release that brings advanced features and 40% less memory footprint to AnyCable—the product to build real-time features for Ruby on Rails apps. The two brand-new PRO features we're presenting in the post, [JWT identification][jwt-id-docs] and [signed streams support][signed-streams-docs], arrived from the Early Access program. We asked its adopters to share how they use _cables_ and which features they would like us to provide out of the box. They might seem unrelated to each other at first glance, but when used together, they have the potential to bring about a huge performance boost.

> When we started collecting applications for the [AnyCable PRO][pro] early access program back in July, we asked respondents to share the way they use _cables_ and which features they would like us to provide out of the box. Analyzing these responses led to the development of two new PRO features, [JWT identification][jwt-id-docs] and [signed streams support][signed-streams-docs]. They might seem unrelated to each other at first glance, but when used together, they have the potential to bring about a huge performance boost. Continue reading to learn more!

Reviewer: to share the way --> to share how

Author: ```suggestion
When we started collecting applications for the [AnyCable PRO][pro] early access program back in July, we asked respondents to share how they use _cables_ and which features they would like us to provide out of the box. Analyzing these responses led to the development of two new PRO features, [JWT identification][jwt-id-docs] and [signed streams support][signed-streams-docs]. They might seem unrelated to each other at first glance, but when used together, they have the potential to bring about a huge performance boost. Continue reading to learn more!
```

> We've just launched the final [AnyCable PRO][pro] release that brings advanced features and 40% less memory footprint to AnyCable—the product to build real-time features for Ruby on Rails apps. The two brand-new PRO features we're presenting in the post, [JWT identification][jwt-id-docs] and [signed streams support][signed-streams-docs], arrived from the Early Access program. We asked its adopters to share how they use _cables_ and which features they would like us to provide out of the box. They might seem unrelated to each other at first glance, but when used together, they have the potential to bring about a huge performance boost.

Reviewer: release that brings advanced features and 40% less memory footprint to AnyCable—the product to build real-time features for Ruby on Rails apps. --> release, bringing advanced features and a 40% lower memory footprint to AnyCable–the product for building real-time features for Ruby on Rails apps.

Reviewer: @palkan

Author: ```suggestion
We've just launched the final [AnyCable PRO][pro] release, bringing advanced features and a 40% lower memory footprint to AnyCable—the product for building real-time features for Ruby on Rails apps. The two brand-new PRO features we're presenting in the post, [JWT identification][jwt-id-docs] and [signed streams support][signed-streams-docs], arrived from the Early Access program. We asked its adopters to share how they use _cables_ and which features they would like us to provide out of the box. They might seem unrelated to each other at first glance, but when used together, they have the potential to bring about a huge performance boost.
```

{intro}

<div class="divider"></div>

## Dealing with authentication

> Amongst all the requested features, one of the leading requests was a "token-based authentication". Most Action Cable applications rely on cookies as an authentication mechanism (either directly or via sessions). The main benefit of cookie-based authentication is simplicity—it just works. However, there are some drawbacks:

Reviewer 2: Again, I think we should explain to newcomers, why we are talking about Action Cable here. I'd like to suggest something like that (but it may be not technically correct so please help me to improve):

AnyCable reckons upon the Action Cable API (although it can work outside of Rails) and most Action Cable applications rely on cookies as an authentication mechanism (either directly or via sessions).

> Amongst all the asked about features, one of the leading requests was a "token-based authentication". Most Action Cable applications rely on cookies as an authentication mechanism (either directly or via sessions). The main benefit of cookie-based authentication is simplicity–it just works. However, there are some drawbacks:

Reviewer: **Sorry, I tried to mix it up but it doesn't work in the final version..**
Amongst all the asked about features, --> Amongst all the requested features,

Author: ```suggestion
Amongst all the requested features, one of the leading requests was a "token-based authentication". Most Action Cable applications rely on cookies as an authentication mechanism (either directly or via sessions). The main benefit of cookie-based authentication is simplicity–it just works. However, there are some drawbacks:
```

> Amongst all the requested features, one of the leading requests was a "token-based authentication". AnyCable reckons upon the Action Cable API (although it can work outside of Rails), and most Action Cable applications rely on cookies as an authentication mechanism (either directly or via sessions). The main benefit of cookie-based authentication is simplicity—it just works. However, there are some drawbacks:

Reviewer: AnyCable reckons --> AnyCable relies 

@palkan it seems this was a mistake

Reviewer 2: It's double "rely" in one sentence then. Could you please suggest a synonym for one of them?

Reviewer: _I would replace the second case:_
most Action Cable applications rely on cookies --> most Action Cable applications utilize cookies

Reviewer 2: Ok!

Author: ```suggestion
Amongst all the requested features, one of the leading requests was a "token-based authentication". AnyCable relies on the Action Cable API (although it can work outside of Rails), and most Action Cable applications utilize cookies as an authentication mechanism (either directly or via sessions). The main benefit of cookie-based authentication is simplicity—it just works. However, there are some drawbacks:
```

- Authenticating non-web clients (e.g., mobile apps) is becoming cumbersome.

> - Rails API-only apps usually do not use cookies for web clients authentication as well (they use tokens).

Reviewer: The phrase "as well" is lost here, so, it doesn't make sense. What would be a synonym? I didn't get it myself. It can't be "too" because it wouldn't fit as a negative sentence.

Author: ```suggestion
- Rails API-only apps usually do not use cookies for web clients authentication (they use tokens).
```

- WebSockets do not offer CORS support and, thus, are **vulnerable to cross-site request forgery** (or [cross-site WebSocket hijacking][cross-site-ws-hijack]).

> Using tokens could certainly help us solve these problems—but what's the catch? Well, we'd have to build everything ourselves, both the server and the client side (and good luck actually finding a library or gem which makes this process easier instead of harder).

Reviewer: both the server and the client side --> both the server, and the client side

Author: ```suggestion
Using tokens could certainly help us solve these problems—but what's the catch? Well, we'd have to build everything ourselves, both the server, and the client side (and good luck actually finding a library or gem which makes this process easier instead of harder).
```

While I was thinking about how to incorporate tokens into AnyCable, I realized that we could also leverage this feature to improve performance. Tokens could also be made to carry **identification** information, and therefore could be used instead of calling an RPC server (`Authenticate` method).

Let's recall how Action Cable (and AnyCable) work in the context of authentication and identification.

Whenever a new connection opens, the `ApplicationCable::Connection#connect` method is called. This is the place where you can reject the connection ([`#reject_unathorized_connection`][reject-connection]) and where you configure the so-called _connection identifiers_ ([`.identifed_by`][ac-identifiers]). The identifiers represent the client state which could be used by channels (e.g., `current_user` or `current_tenant`). AnyCable obtains the identifiers during the `Authenticate` call and passes along all subsequent requests (in a [serialized form][identifiers-serialization]).

> To sum everything up, all we need from the `#connect` method is to accept-or-reject the connection and populate the identifiers. And we can encapsulate all this logic within a single [JWT][] token!

Reviewer: Is there a reason that it's written like "accept-or-reject" instead of "accept or reject"?

Author: Not really.

```suggestion
To sum everything up, all we need from the `#connect` method is to accept or reject the connection and populate the identifiers. And we can encapsulate all this logic within a single [JWT][] token!
```

      # and is responsible for converting strings back to objects
      identifiers = deserialize_identifiers(JSON.parse(decoded_token["ext"]))

      # Populate identifiers
      identifiers.each do |k, v|
        self.public_send("#{k}=", v)
      end
    rescue JWT::DecodeError
      reject_unauthorized_connection
    end
  end
end
```

This doesn't look like such a simple solution, after all 😕 If you continue staring at this piece of code for a bit longer, you might notice that this implementation contains nothing specific to a particular application: we retrieve a token, verify it, and extract the identifiers. If this is the case, why can't we just re-implement this in Go? Well, we can!

This is how AnyCable Go PRO JWT identification works:

- A client passes a token along with the connection request (via a query string or a header).
- A Go server retrieves a header, verifies its signature and TTL, and rejects the connection if token is invalid.
- If the token is valid, we extract the identifiers, store them in the connection metadata and finalize the handshake (by sending a `{"type":"welcome"}` message).

> And we do not perform any RPC calls! Let's see how it affects the connection time:

Reviewer: do not --> don't have to

Author: ```suggestion
And we don't have to perform any RPC calls! Let's see how it affects the connection time:
```

<figure class="blog--figure">
<video loop="loop" muted="muted" autoplay="autoplay" playsinline="true" width="1221" height="800" class="blog--media">
  <source src="./connect-benchmark.h264.mp4" ttype="video/mp4; codecs=avc1.4D401E,mp4a.40.2">
</video>
</figure>

It's more than 2x faster now! And even more importantly, we've reduced the stress on the RPC server. This could be especially useful during _connection avalanches_.

> JWT identification allows you to standardize the authentication flow for WebSockets, protects from cross-site WebSocket hijacking and results in a performance boost!

> Luckily, you don't need to write any of the code above yourself, even the token generation part a little companion gem, [anycable-rails-jwt][], which does this for you!

Reviewer: **Current text has an error of grammatical structure, change to:**

Luckily, you don't need to write any of the code above yourself, even the token generation part is done for you via a little companion gem called [anycable-rails-jwt][].

Author: ```suggestion
Luckily, you don't need to write any of the code above yourself, even the token generation part is done for you via a little companion gem called [anycable-rails-jwt][].
```

> The only question left is how to handle tokens expiration gracefully? AnyCable Go uses a specific disconnect reason to distinguish expiration from unauthorized access (sends a `{"type":"disconnect","reason":"token_expired"}` message). You can use it to refresh the token. Again, we're glad to provide an out-of-the-box solution for that in our brand new [anycable-client-refresh-tokens][]!

Reviewer: **IMO, looks quite odd to ask a question as it's currently being done, my suggestion:**
The only question left is how to handle tokens expiration gracefully? --> "The only question left: how can we gracefully handle token expiration?"
**Next suggestion:**
AnyCable Go users a. specific disconnect reason --> AnyCable Go uses a specific disconnect message
**Third suggestion:**
solution for that --> solution for this

> The only question left is how to handle tokens expiration gracefully? AnyCable Go uses a specific disconnect reason to distinguish expiration from unauthorized access (sends a `{"type":"disconnect","reason":"token_expired"}` message). You can use it to refresh the token. Again, we're glad to provide an out-of-the-box solution for that in our brand new [anycable-client-refresh-tokens][]!

Reviewer: **IMO, looks quite odd to ask a question as it's currently being done, my suggestion:**
The only question left is how to handle tokens expiration gracefully? --> "The only question left: how can we gracefully handle token expiration?"
**Next suggestion:**
AnyCable Go users a. specific disconnect reason --> AnyCable Go uses a specific disconnect message
**Third suggestion:**
solution for that --> solution for this

Author: ```suggestion
The only question left: how can we gracefully handle token expiration? AnyCable Go uses a specific disconnect message to distinguish expiration from unauthorized access (sends a `{"type":"disconnect","reason":"token_expired"}` message). You can use it to refresh the token. Again, we're glad to provide an out-of-the-box solution for this in our brand new [anycable-client-refresh-tokens][]!
```

Let's start with the [turbo-rails][] gem, which integrates Action Cable with Hotwire. You just need to drop a helper into your template to start consuming streams:

```erb
<%= turbo_stream_from some_model %>
```

The helper adds a [`<turbo-cable-stream-source>` element][turbo-cable-stream-source], which initiates an Action Cable subscription. The corresponding server-side code is as follows:

```ruby
class Turbo::StreamsChannel < ActionCable::Channel::Base
  extend Turbo::Streams::Broadcasts, Turbo::Streams::StreamName

  def subscribed
    if verified_stream_name = self.class.verified_stream_name(params[:signed_stream_name])
      stream_from verified_stream_name
    else
      reject
    end
  end
end
```

The `#verified_stream_name` method uses an [`ActiveSupport::MessageVerifier`][message-verifier] to decode and verify the stream name. The subscription is rejected if the signed stream name is invalid.

Structurally speaking, the code above is the same as the example we wrote for JWT identification. What does this mean? It means that, **once again**, we can "move" the implementation to an AnyCable Go server and **create subscriptions without touching RPC**.

> This is exactly what we did as a part of the [signed streams][signed-streams-docs] feature. Further, we did it not only for Turbo Streams, but for Cable Ready as well (since its [`#stream_from` implementation][cr-stream-from] is nearly the same).

Author: ```suggestion
This is exactly what we did as a part of the [signed streams][signed-streams-docs] feature. Further, we did it not only for Turbo Streams, but for CableReady as well (since its [`#stream_from` implementation][cr-stream-from] is nearly the same).
```

> > By combining JWT identification with signed streams, it's possible to completely avoid running an RPC server, in case you only need to use Hotwire or CableReady functionality.

Reviewer: _There is an inconsistency here. In the line above it is called "Cable Ready" but here it is called "CableReady"_ @palkan

Below is a visualization of RPC server metrics during Hotwire benchmarks with four different configurations (denoted by blue markers), from left to right: baseline (AnyCable PRO without any features enabled), with JWT identification, with signed streams, and with both features enabled at the same time.

<figure class="blog--figure">
  <img class="blog--media" title="AnyCable metrics when running with and without JWT and 'hot' streams" src="./grafana.png" width="1364" height="624">
</figure>

It shouldn't come as any surprise that the last one shows zeros for RPC metrics—it wasn't impacted at all. We can also note that CPU usage with AnyCable Go is the lowest in the fourth scenario—**dealing with tokens and signatures is "cheaper" than performing gRPC calls**.

<div class="divider"></div>

> We've added these features in response to our users needs, and because we believe that we're able to build a much better product by working together. Want to join the team? Give [AnyCable a try today](/)!

Reviewer: QUESTION: Are we talking about paid features in this article? (The PRO version is not free, correct?)

Author: I'm not an experienced pitch-writer, so, I'd prefer to leave it up to you 🙂; WDYT?

Reviewer: **Here's my suggestion, at least, this can be a starting point:**
We've added these features in response to your needs. We believe that, thanks to user feedback, we're able to build a much better product together. Who knows? Maybe your request will be the next feature we add! Ready to get started? Give [AnyCable a try today](/)!

Author: ```suggestion
We've added these features in response to your needs. We believe that, thanks to user feedback, we're able to build a much better product together. Who knows? Maybe your request will be the next feature we add! Ready to get started? Give [AnyCable a try today](/)!
```
